zebra-state/src/service.rs

```diff
@@ -226,16 +234,17 @@ impl StateService {
         prepared: &PreparedBlock,
     ) -> Result<(), ValidateContextError> {
         let relevant_chain = self.any_ancestor_blocks(prepared.block.header.previous_block_hash);
-        assert!(relevant_chain.len() >= POW_AVERAGING_WINDOW + POW_MEDIAN_BLOCK_SPAN,
-                "contextual validation requires at least 28 (POW_AVERAGING_WINDOW + POW_MEDIAN_BLOCK_SPAN) blocks");
 
+        // Security: check proof of work before any other checks
         check::block_is_contextually_valid(
             prepared,
             self.network,
             self.disk.finalized_tip_height(),
             relevant_chain,
         )?;
 
+        check::nullifier::no_duplicates_in_finalized_chain(prepared, &self.disk)?;
+
```

Comment:
I'm trying to work out the best way to test both the non-finalized and finalized state code together.

Moving the assertions works, and it's a simple change, but it does separate the assertion from the code that requires it.

I could move `check::nullifier::no_duplicates_in_finalized_chain` and `self.mem.commit...` into a method that is called by `validate_and_commit`. But I'm not sure if that solution will work for other similar contextual checks.

Maybe we can see how it goes as we implement the other state designs?

Reply:
I had a think about this - we've already checked the difficulty and bound the transaction IDs to the block header.

So any future tests will just have to fix up the authorizing data commitment and chain history commitment. Which seems fine. We can make a test-only method for that - the chain history tests will need it anyway.